getting_started/scripting/c_sharp/c_sharp_style_guide.rst

```diff
@@ -0,0 +1,299 @@
+                        break;
+                }
+            }
+
+            i += (int)MyProperty; // No space after a type cast.
+        }
+    }
+
+Naming Conventions
+------------------
+
+Use *PascalCase* for all namespaces, type names and member level identifiers (i.e. methods, properties,
+constants, events), except for private fields:
+
+.. code-block:: csharp
+
+    namespace ExampleProject
+    {
+        public class PlayerCharacter
+        {
+            public const float DefaultSpeed = 10f;
+
+            public float CurrentSpeed { get; set; }
+
+            protected int HitPoints;
+
+            private void CalculateWeaponDamage()
+            {
+            }
+        }
+    }
+
+Use *camelCase* for all other identifiers (i.e. local variables, method arguments), and use
+underscore('_') as a prefix for private fields (but not for methods or properties, as explained above):
+
+.. code-block:: csharp
+
+    private Vector3 _aimingAt; // Use '_' prefix for private fields.
+
+    private void Attack(float attackStrength)
+    {
+        Enemy targetFound = FindTarget(_aimingAt);
+
+        targetFound?.Hit(attackStrength);
+    }
+
+There's an exception with acronyms which consist of two letters like *'UI'* which should be written in
+upper case letters when used where Pascal case would be expected, and in lower case letters otherwise.
+
+Note that *'id'* is **not** an acronym, so it should be treated as a normal identifier:
+
+.. code-block:: csharp
+
+    public string Id { get; }
+
+    public UIManager UI
+    {
+        get { return uiManager; }
+    }
+
+It is generally discouraged to use a type name as a prefix of an identifier like *'string strText'*
+or *'float fPower'*, for example. However, there's an exception about interfaces, in which case they
+**should** be named using an upper case *'I'* as a prefix, like *'IInventoryHolder'* or *'IDamageable'*.
+
+Lastly, consider choosing descriptive names and do not try to shorten them too much if it affects
+readability.
+
+For instance, if you want to write a code to find a nearby enemy and hit with an weapon, prefer
+
+.. code-block:: csharp
+
+    FindNearbyEnemy()?.Damage(weaponDamage);
+
+Rather than,
+
+.. code-block:: csharp
+
+    FindNode()?.Change(wpnDmg);
+
+Implicitly Typed Local Variables
+--------------------------------
+
+Consider using implicitly typing (*'var'*) for declaration of a local variable, but do so
+**only when the type is evident** from the right side of the assignment:
+
+.. code-block:: csharp
+
+    // You can use `var` for these cases:
+
+    var direction = new Vector2(1, 0);
+
+    var value = (int)speed;
+
+    // But not for these:
+
+    var value = GetValue();
+
+    var speed = 1;
+
+    var text = "Some value";
+
+    for (var i = 0; i < 10; i++)
+    {
+    }
+
+Other Considerations
+--------------------
+
+ * Use explicit access modifiers.
+ * Use properties instead of non-private fields.
```

Comment:
Why? And does this mean we should change the fields such as `real_t x` in `Vector2` to a property?

I don't really see any reason `public real_t x;` is bad. It should always be changeable from anywhere, right?

Reply:
@aaronfranke I believe it's an old convention shared with Java devs, and the most important reason could be that if you expose something as a property, you can always change the underlying implementation (i.e. adding validation or returning a calculated value) later without affecting existing users.

In that particular case with `Vector2`, I agree that using public fields won't likely to cause any practical trouble.

But I don't see any reason not use properties in that case either, so I think it's worth to include it in the guide. If we don't want to change exising code in `Vector2` and other similar cases, maybe we can change the phrase to read it more like a recommendation rather than a hard and fast rule.

Reply:
I just found a relevant comment in the C# guide from MS : 
> Generally, you should use fields only for variables that have private or protected accessibility. Data that your class exposes to client code should be provided through methods, properties and indexers. By using these constructs for indirect access to internal fields, you can guard against invalid input values.

* https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/fields

On a side note, it's interesting to see that it makes an exception for protected fields also. Personally, I've always assumed protected members to be a part of the API reference so I wonder if the practice differs among C# developers.

Reply:
Typically wouldn't you know in advance if the field/property would need the ability to have a differently coded implementation? IMO it'd be overcomplicated/possibly confusing to use properties everywhere, even when a field would do.

Reply:
In general, yes. But probably we shouldn't shut out such possibilities of replacing such fields with some native methods, or adding extra codes like cache flushing or validation to them later. 

It also allows extracting an interface later, and can make the code more consistent when there are a few of such members that need to be restricted for update.

I don't have too strong an opinion about changing existing fields in such types like `Vector2`, but I think it's a good idea to include it among other recommendations.

Reply:
How about we make the text "Consider using ..."?

Reply:
It sounds reasonable to me, especially if we are not going to change all such instances in existing code. I'd like to hear other people's thoughts before proceeding though.

Reply:
"Prefer properties over public fields."?

Reply:
I think that there is value in having a consistent API.

Once we have finished discussion on this style guide I hope that it is applied consistently across the public API before 3.1.

With that in mind we will have to revisit the example `real_t x`, at least to change it to `real_t X`.  It doesn't seem like much more work to make it `real_t X { get; set; }` while we are at it.

The public API should be binding once final as we won't want to break people's projects.  This is a classic use case for properties over fields.

Reply:
No, use properties instead of public or protected instance fields. The reasons for using properties over fields are all over the internet, the main one being encapsulation.

Now, if you have a good reason or no other choice, go on. There are exceptions for every rule.

Reply:
Noteworthy: Never using public fields means that `camelCase` without underscores will only be used for local variables and method arguments.

Reply:
So, should I change the relevant line ("Use properties instead of non-private fields.") to a more suggestive form, or I better leave it as it is?

I'll update the PR as soon as it becomes clear on this one.

Reply:
@mysticfall I'd let neikeq make the call, so we can leave as is or change to the more explicit "Use properties instead of public or protected instance fields." instead. Shall I merge this, then? We can keep editing it forever anyway, but it's now been up for a week to review and all interested parties seem to have had a glance at it.

Reply:
I would absolutely be fine with merging as-is.